### ISSUE TITLE
Image intensity augmentations

### ISSUE THREAD

**Issue by IDoCodingStuffs** [TorchIO-project/torchio#1187]
**🚀 Feature**
Transforms that shift voxel intensity, such as intensity flipping (i.e. 1 - val, val in [0, 1]), cluster and remap, contrast jitter etc.

**Motivation**
I am working on a spine segmentation problem on MRI images where I need to train a model to perform in multiple pulse sequence modalities, but training data has only a single modality. As such, models tend to pick up on intensity features and perform very poorly on different modalities with different intensity distributions. (see the middle image vs others)

![image](https://github.com/user-attachments/assets/daef2e75-fed8-48b1-b81f-e23298966398)

**Pitch**
Adding transforms to shift intensity features would allow models to pick up on shapes and contours rather than learning intensity values as features.

**Alternatives**
One very basic approach could be something like this
```
class RandomFlipIntensity:
    def __call__(self, input: tio.Subject, p=0.5):
        if np.random.random() <= p:
            max_intensity = torch.max(input["image"].tensor)
            min_intensity = torch.min(input["image"].tensor)
            flipped_intensity = max_intensity - input["image"].tensor + min_intensity
            input = tio.Subject(
                image=tio.ScalarImage(tensor=flipped_intensity),
                segmentation=tio.LabelMap(tensor=input["segmentation"].tensor),
            )
        return input
```

**Comment by romainVala** [comment]
Hi 

Transforming intensity in a physical plausible manner, is not easy, and there are very few effective transform ( the only one in torchio was RandomGamma, but from my experience,  It was not effective enough) . The transform you propose sound good because it makes a radical change by inverting the  intensity ... so ...worth to add 

I would be very interested to know, if this is enough to generalize properly.

The best approach to get a contrast agnostic model, is Billot SynthSeg proposition, since it trains the model with really random contrast, and the results are impressive (it does segment any contrast !)
but one need very good realistic label (including structure in the background), so it may not be easily applicable for your use case

**Comment by Jessy-Ding** [comment]
Hi — this looks unassigned with no open PR, so I'd like to pick it up if that's OK.

Building on @romainVala's note, I'd suggest scoping the first PR to just that — a `RandomIntensityInversion`
transform — and leaving the cluster-and-remap / contrast-jitter ideas from the
original request as separate follow-ups. 

I've prototyped it following the existing `RandomGamma`/`Gamma` layout (a
deterministic `IntensityInversion` + a `RandomIntensityInversion` wrapper), using
@IDoCodingStuffs's `max - x + min`, since it doesn't
assume `[0, 1]` normalized inputs and works directly on raw MRI intensities. The
operation is its own inverse, so it should fit the invertibility framework naturally.

Two design questions before I open a PR:

1. **min/max scope** — the prototype takes a single min/max over the whole tensor,
   while the other intensity transforms here operate per-channel. Per-channel seems
   more correct for multi-channel inputs (identical for single-channel MRI). Any
   preference?
2. **Naming** — `RandomIntensityInversion` / `IntensityInversion`, or shorter?

On @romainVala's open question of whether inversion alone is enough to generalize —
this gives a clean, mergeable building block to test that on real cross-modality
data, and a SynthSeg-style random-contrast transform could be a heavier follow-up
if there's appetite.

Thanks!

**Comment by romainVala** [comment]
hi
1)
yes when dealing with intensity transform, it make sense to defined the parameters channel by channel 

2) About the naming, I will choose the more relevent. Is your transform random ? or is it a "binairy states" ie with or without inversion ?

**Comment by Jessy-Ding** [comment]
> hi 1) yes when dealing with intensity transform, it make sense to defined the parameters channel by channel
> 
> 2. About the naming, I will choose the more relevent. Is your transform random ? or is it a "binairy states" ie with or without inversion ?

Hi @romainVala, thanks!

1. Great, I'll compute the parameters per channel.

2. Good question — it's a binary state, not a parametric random transform. The inversion itself is fully deterministic: for a given image it's just `max - x + min` (per channel), so there's no continuous parameter to sample. The only stochastic part is whether the inversion is applied or not, i.e. an identity-or-inverted behavior with some probability `p`. So it's not "random" in the same sense as `RandomGamma`, which samples a value from a range. I'll leave the naming to you since you'd prefer to pick the most relevant one — I just wanted to confirm the semantics so it's clear it's a two-state transform rather than a parametrized one.

**Comment by romainVala** [comment]
exactly, that is why I would call it IntensityInversion and put it in  the preprocessing/Intensity transform

**Comment by fepegar** [comment]
Hi all.

@IDoCodingStuffs: apologies, I had missed this issue.
@Jessy-Ding are you a coding agent?
@romainVala I think before adding this to the library I'd like to know whether it helps at all.

**Comment by Jessy-Ding** [comment]
> Hi all.
> 
> [@IDoCodingStuffs](https://github.com/IDoCodingStuffs): apologies, I had missed this issue. [@Jessy-Ding](https://github.com/Jessy-Ding) are you a coding agent? [@romainVala](https://github.com/romainVala) I think before adding this to the library I'd like to know whether it helps at all.

Hi, @fepegar, no, I'm an MSCS student at Northeastern and am currently taking an open-source course at CodePath that teaches me how to contribute to open-source repos. I have years of experience in medical imaging and MRI scans. Of course, I use AI to help draft plans and code, too, and I'm currently looking into these issues and trying to discuss ways to solve them.

**Comment by Jessy-Ding** [comment]
I'll simplify the v2 code: use a single deterministic IntensityInversion in transforms/intensity/ without a Random* wrapper. Since v2 merged Random/non-Random (like RandomGamma into Gamma) and the Transform class provides p, separate RandomIntensityInversion adds no benefit. IntensityInversion(p=0.5) naturally inverts half the time. Min/max is per-channel; on/off is per-image. The op's inverse nature aligns with invertibility. @fepegar @romainVala

**Comment by fepegar** [comment]
Thanks for clarifying, @Jessy-Ding. As I said, I'm not sure how useful this transform is, particularly for CT, and before it's added to the library I'd like to see some evidence that it improves results in an application. If that is something you can demonstrate, we can plan the addition of the transform to the library.

**Comment by IDoCodingStuffs** [comment]
Would we want to set up a basic experiment to demonstrate usefulness? i.e. run training with vs without and compare on some toy dataset/problem

**Comment by fepegar** [comment]
Yes, I think that would make sense.

**Comment by Jessy-Ding** [comment]
<p><strong>Efficacy evidence for <code>IntensityInversion</code> (synthetic toy, CPU, no downloads)</strong></p>
Hi, dear maintainers, @fepegar @IDoCodingStuffs @romainVala 
<p>I ran the basic train-with-vs-without experiment requested in the thread.</p>
<p><strong>Setup.</strong> Procedurally generated 2D images of random shapes; task = segment the shapes (Dice). Training images always have <strong>bright foreground on dark background</strong>. Two test sets: an <strong>in-domain</strong> set (same polarity) and an <strong>OOD</strong> set with <strong>inverted contrast</strong> (foreground darker than background). The OOD set is built with a <em>different</em> non-linear intensity-decreasing map than the trained <code>max − x + min</code>, so a gain reflects learned polarity-invariance rather than memorizing the augmentation. Only one variable changes between arms:</p>
<ul>
<li><strong>Arm A:</strong> flip / affine / noise</li>
<li><strong>Arm B:</strong> Arm A + <code>IntensityInversion(p=0.5)</code></li>
</ul>
<p>3 seeds, seed-paired.</p>
<p><strong>Results (mean ± std, Dice):</strong></p>

Test set | Arm A | Arm B | Δ
-- | -- | -- | --
OOD (inverted contrast) | 0.042 ± 0.018 | 0.835 ± 0.040 | +0.793
In-domain (same polarity) | 0.996 ± 0.001 | 0.978 ± 0.005 | −0.018


<p>The OOD gain is large and consistent across all 3 seeds, while in-domain performance is essentially unchanged (within the pre-registered &#x3C;2-point guard). The figure shows the effect: on an inverted-contrast image, the baseline segments the bright background (it learned "bright = foreground"), while the <code>IntensityInversion</code> model correctly segments the shapes.</p>
<p><strong>Scope:</strong> this is an MRI-style cross-contrast augmentation. CT intensities (Hounsfield units) are physically meaningful, so I do <strong>not</strong> claim benefit for CT.</p>

<img width="1080" height="360" alt="Image" src="https://github.com/user-attachments/assets/fffc7ced-073b-4e61-a591-412750b07289" />

**Comment by fepegar** [comment]
Thank you for running that, @Jessy-Ding. Unfortunately I don't think the results with the toy data are significant enough to move forward. I would like to see results on a real dataset and task.

**Comment by Jessy-Ding** [comment]
<p>Update @fepegar — Here I am reporting the full picture, including a negative result.</p>
<p><code>IntensityInversion</code> reliably helps only when the train→test shift is (close to) a global intensity-polarity flip. On real cross-modality MRI — even in a setup matched to the original use case — I could not find a reliable benefit. Your skepticism looks justified.</p>
<p><strong>1. Synthetic toy (controlled polarity flip).</strong> Train on bright-foreground shapes, test on inverted-contrast shapes (a genuine polarity flip), segmentation. Inversion helped strongly and consistently: OOD Dice <strong>0.04 → 0.84</strong> (+0.79), all seeds, ~0 in-domain cost. So the transform does what it claims <em>when the shift is actually a polarity flip</em>.</p>
<p><strong>2. BraTS cross-modality (mis-designed).</strong> Train on one MRI sequence, test on another, whole-tumor segmentation. This was the wrong test: whole-tumor (edema) is bright on FLAIR/T2 but <strong>nearly absent on T1</strong>, so the target <em>vanishes</em> across modalities (OOD Dice ≈0.12 for both arms — no signal for any augmentation to recover). It does not represent the requester's scenario, where the target is a structure visible in <em>every</em> modality.</p>
<p><strong>3. CHAOS cross-modality (use-case-matched).</strong> This matches the requester's situation: an anatomical structure visible in all modalities, only the intensity distribution changes. Liver segmentation, train on <strong>T1 (in-phase)</strong>, test on <strong>T2-SPIR</strong> (held-out subjects), <code>IntensityInversion(p=0.5)</code> the only added transform, Dice-only loss, <strong>8 seeds</strong>.</p>

Test set | Arm A (baseline) | Arm B (+inversion) | Δ
-- | -- | -- | --
OOD (T2) | 0.211 ± 0.069 | 0.228 ± 0.062 | +0.017
in-domain (T1) | 0.698 ± 0.022 | 0.667 ± 0.021 | −0.032


<p>Per-seed ΔOOD: <code>+0.057, +0.164, −0.018, −0.147, +0.048, −0.005, −0.103, +0.140</code> → <strong>4 up / 4 down</strong>, mean +0.017 with per-seed swings of ±0.15. The gain is within noise. Meanwhile in-domain costs a consistent ~0.03.</p>
<p>Real multi-sequence MRI intensity differences aren't a global polarity inversion — they're nonlinear, tissue-specific, and change the whole histogram. Inversion addresses only the polarity-flip component, which isn't the dominant mode of real cross-modality variation, so it doesn't reliably help cross-modality transfer here.</p>

**Comment by romainVala** [comment]
thanks for reporting you test, 
and I agree with your conclusion on real data MRI contrast difference are fare from the simple linear, so it does not help much. 

@Jessy-Ding you mention for your toy test case construction : ` The OOD set is built with a different non-linear intensity-decreasing map than the trained max − x + min`

this may be an idea to improve the transform -> adding more random non linear intensity deformation may help

(also  testing with large random gamma)



@fepegar: not all transform are useful ... and to test, one need to have it implemented ... my feeling is that the current stat of torchio's intensity random transform is not changing enough the intensity distribution (since ODD is often an issue), I guess  here are place for improvement

**Comment by Jessy-Ding** [comment]
Hi. @romainVala @fepegar <p>Quick update after digging into this properly. I wanted to actually test whether intensity augmentation helps the cross-sequence case from the issue (train on one MRI contrast, run on another), so I ran a real-data study on CHAOS (abdominal MR, train on T1, test on held-out T2, 4 organs, a few seeds). Same model/data/seeds across arms — only the augmentation changes.</p>
<p>Cross-modality Dice gain over a plain-augmentation baseline (mean across the 4 organs):</p>

augmentation | ΔDice on the unseen contrast
-- | --
IntensityInversion alone | +0.01 (helped on 3/4 organs)
random non-linear remap (GIN-style conv) | +0.04 (4/4)
full intensity suite (gamma + bias field + inversion + GIN) | +0.13 (4/4)
synthesize images from the labels (SynthSeg-style) | +0.31 (4/4)


<p>So the honest ladder: any single intensity transform (inversion included) does very little on its own. Stacking a bunch of strong ones (domain randomization) helps a fair bit. But the thing that actually moves the needle is not augmenting the real image at all — it's generating training images from the label maps with random contrast, i.e. the SynthSeg approach. It roughly doubles the gain of the whole suite, and it's biggest exactly where the baseline fails worst (small organs that are nearly invisible after the contrast switch).</p>
<p>The useful part: <strong>TorchIO already supports this.</strong> <code>LabelsToImage</code> is the SynthSeg synthesis primitive (its own docstring even says so) — compose it with <code>Blur</code> + <code>BiasField</code> + noise and you've got contrast-agnostic training, no extra code. I ran the SynthSeg arm above with <code>tio.LabelsToImage</code> directly and got the same result as my hand-rolled version, so this isn't hypothetical.</p>
<p>For the original spine question specifically, the relevant prior work:</p>
<ul>
<li><strong>SynthSeg</strong> (Billot et al., MedIA 2023) — train on synthetic random-contrast images from labels; segments any contrast without retraining.</li>
<li><strong>Contrast-agnostic spinal cord segmentation</strong> (Bédard et al., MedIA 2025) — exactly this problem, already shipped in Spinal Cord Toolbox (<code>sct_deepseg</code>); single model, aggressive aug + soft labels. Probably usable off the shelf.</li>
<li>The single-source DG augmentation family this all sits in: <strong>BigAug</strong> (Zhang 2020), <strong>GIN</strong> (Ouyang, IEEE-TMI 2022/2023), <strong>RandConv</strong> (Xu, ICLR 2021).</li>
</ul>
<p>One question: would a short doc/example showing how to wire <code>LabelsToImage</code> + <code>Blur</code> + <code>BiasField</code> for contrast-agnostic training be useful? If you don't think that's needed, then there's basically no new code to add here.</p>

**Comment by fepegar** [comment]
Ok I think we can use our time to fix real issues in the new version instead of discussing with a bot whether to add a transform of questionable usefulness for some coursework.
